src/charm.py

```diff
@@ -224,37 +220,63 @@ def _on_database_storage_detaching(self, _) -> None:
 
-    def _on_get_cluster_admin_credentials(self, event: ActionEvent) -> None:
-        """Action used to retrieve the cluster admin credentials."""
-        event.set_results(
-            {
-                "cluster-admin-username": CLUSTER_ADMIN_USERNAME,
-                "cluster-admin-password": self._get_secret("app", CLUSTER_ADMIN_PASSWORD_KEY),
-            }
-        )
+    def _on_get_password(self, event: ActionEvent) -> None:
+        """Action used to retrieve the system user's password."""
+        username = ROOT_USERNAME
 
-    def _on_get_server_config_credentials(self, event: ActionEvent) -> None:
-        """Action used to retrieve the server config credentials."""
-        event.set_results(
-            {
-                "server-config-username": SERVER_CONFIG_USERNAME,
-                "server-config-password": self._get_secret("app", SERVER_CONFIG_PASSWORD_KEY),
-            }
-        )
+        if "username" in event.params:
+            username = event.params["username"]
 
-    def _on_get_root_credentials(self, event: ActionEvent) -> None:
-        """Action used to retrieve the root credentials."""
-        event.set_results(
-            {
-                "root-username": "root",
-                "root-password": self._get_secret("app", ROOT_PASSWORD_KEY),
-            }
-        )
+        if username not in REQUIRED_USERNAMES:
+            raise RuntimeError("Invalid username.")
 
-    def _get_cluster_status(self, event: ActionEvent) -> None:
-        """Action used to retrieve the cluster status."""
-        event.set_results(self._mysql.get_cluster_status())
+        if username == ROOT_USERNAME:
+            secret_key = ROOT_PASSWORD_KEY
+        elif username == SERVER_CONFIG_USERNAME:
+            secret_key = SERVER_CONFIG_PASSWORD_KEY
+        elif username == CLUSTER_ADMIN_USERNAME:
+            secret_key = CLUSTER_ADMIN_PASSWORD_KEY
+        else:
+            raise RuntimeError("Invalid username.")
+
+        event.set_results({"username": username, "password": self._get_secret("app", secret_key)})
+
+    def _on_set_password(self, event: ActionEvent) -> None:
+        """Action used to update/rotate the system user's password."""
+        if not self.unit.is_leader():
+            raise RuntimeError("set-password action can only be run on the leader unit.")
+
+        username = ROOT_USERNAME
+
+        if "username" in event.params:
+            username = event.params["username"]
+
+        if username not in REQUIRED_USERNAMES:
+            raise RuntimeError("Invalid username.")
+
+        if username == ROOT_USERNAME:
+            secret_key = ROOT_PASSWORD_KEY
+        elif username == SERVER_CONFIG_USERNAME:
+            secret_key = SERVER_CONFIG_PASSWORD_KEY
+        elif username == CLUSTER_ADMIN_USERNAME:
+            secret_key = CLUSTER_ADMIN_PASSWORD_KEY
+        else:
+            raise RuntimeError("Invalid username.")
+
+        new_password = None
```

Comment:
you can replace these lines with this:

```suggestion
        new_password = event.params.get("password") or generate_random_password(PASSWORD_LENGTH)
```

Reply:
I think it will throw KeyError: 'password'